### ISSUE TITLE
NaN car velocity

### ISSUE THREAD

**Issue by benjamincburns** [ZealanL/RocketSim#6]
In this very contrived example (basically just a trace from an attempted training run) we observe that the car velocity is sometimes NaN.

I suppose this could be an issue in the python bindings, but it seems more likely a problem with the underlying engine, hence why I'm reporting it here.

Repro:

```python
from rocketsim import Vec3, Angle
from rocketsim.sim import Arena, CarConfig, GameMode, Team, Car, CarControls, Ball

arena = Arena(GameMode.Soccar)
agent_id = arena.add_car(Team.Blue, CarConfig.Octane)


ball = arena.get_ball()

ball.pos = Vec3(0, 0, 5000)
ball.angvel = Vec3(0, 0, 0)
ball.vel = Vec3(0, 0, 0)

arena.ball = ball

agent = arena.get_car(agent_id)

# x = +- 3000
# y = +- 3000
agent.pos = Vec3(2040.60498046875, -2534.492919921875, 17.0)
agent.angles = Angle(-0.0, 0, 0)
agent.angvel = Vec3(0, 0, 0)
agent.boost = 100

control_dicts = [
    { "throttle": 0.0,  "steer": 0.5,  "pitch": 0.5,  "yaw": 0.5, "roll": 1.0, "jump": False, "boost": False, "handbrake": False },
    { "throttle": -1.0, "steer": 1.0,  "pitch": 1.0,  "yaw": 0.0, "roll": 0.0, "jump": False, "boost": False, "handbrake": True },
    { "throttle": 0.0,  "steer": 0.0,  "pitch": 0.0,  "yaw": 0.75, "roll": 1.0, "jump": True, "boost": False, "handbrake": True },
    { "throttle": 0.0,  "steer": 0.0,  "pitch": 0.0,  "yaw": 0.75, "roll": 0.0, "jump": False, "boost": False, "handbrake": False },
    { "throttle": 1.0,  "steer": -0.5, "pitch": -0.5, "yaw": 0.0, "roll": 0.0, "jump": False, "boost": True, "handbrake": False },
    { "throttle": 0.0,  "steer": 0.5,  "pitch": 0.5,  "yaw": -0.5, "roll": 0.0, "jump": False, "boost": False, "handbrake": False },
    { "throttle": -1.0, "steer": 1.0,  "pitch": 1.0,  "yaw": 0.0, "roll": 0.0, "jump": False, "boost": False, "handbrake": True },
    { "throttle": 0.0,  "steer": 0.0,  "pitch": 0.0,  "yaw": 1.0, "roll": -1.0, "jump": True, "boost": False, "handbrake": True }
]

def control(**kwargs):
    controls = CarControls(**kwargs)
    arena.set_car_controls(agent_id, controls)
    arena.step(8)
    car_vel = arena.get_car(agent_id).get_vel()
    print(f"Car vel: {car_vel}")

[ control(**control_dict) for control_dict in control_dicts ]
````

Output (notice final line has NaN values for `x` and `y`:
```
Car vel: Vec3 { x: -0.07, y: 0.03, z: 137.92 }
Car vel: Vec3 { x: -79.549995, y: -18.42, z: 72.88 }
Car vel: Vec3 { x: -92.6919, y: -13.390303, z: 398.83078 }
Car vel: Vec3 { x: -92.68999, y: -13.389998, z: 355.47 }
Car vel: Vec3 { x: -88.259995, y: -13.59, z: 312.18 }
Car vel: Vec3 { x: -88.259995, y: -13.59, z: 268.82 }
Car vel: Vec3 { x: -92.68999, y: -13.389998, z: 225.48997 }
Car vel: Vec3 { x: NaN, y: NaN, z: 220.06998 }
```

**Comment by benjamincburns** [comment]
@NeverCast says

> It does not reproduce if I set the pos, vel, boost, ang and angvel of the agent and step the last action
> So there is probably some internal state that makes the difference

**Comment by nevercast** [comment]
Angular Velocity is all NaNs in the final state as well.

| Parameter  | Value |
| ------------- | ------------- |
| Position X  | -33.79001  |
| Position Y  | -5.33  |
| Position Z  | 122.57  |
| Velocity X  | NaN  |
| Velocity Y  | NaN  |
| Velocity Z  | 220.07  |
| Angle Pitch | -0.03143  |
| Angle Yaw   | -0.0454  |
| Angle Roll  | -0.149  |
| Angular Velocity X | NaN  |
| Angular Velocity Y | NaN  |
| Angular Velocity Z | NaN  |
| Boost Amount | 0.0  |

**Comment by JPK314** [comment]
This is an issue with stalling, when it sees a flip and a stall is performed the dodgeDir vector is normalized but is zero

**Comment by ZealanL** [comment]
Stalling works fine for me in testing.

I also tested your exact case with the following code:
```c
for (int j = 0; j < 120; j++) {

	arena.Step(1);

	RS_LOG("Tick [" << j << "]:");
	CarState state = car->GetState();
	if (state.isOnGround) {
		RS_LOG(" > Jumping on ground");
		car->controls.jump = !car->controls.jump;
	} else if (state.isJumping) {
		RS_LOG(" > Holding jump")
		car->controls.jump = 1;
	} else if (!state.hasFlipped && !state.hasDoubleJumped) {
		RS_LOG(" > Flipping...");
		car->controls.jump = !car->controls.jump;
		car->controls.yaw = 1;
		car->controls.roll = -1;
		car->controls.handbrake = 1;
	}
	RS_LOG(" > Pos: " << state.pos);
	RS_LOG(" > Vel: " << state.vel);
	RS_LOG(" > AngVel: " << state.angVel);
}
```

The output showed no NANs whatsoever.

![image](https://user-images.githubusercontent.com/36944229/218600137-59569dfe-7847-437c-a70f-8831e364ddec.png)

Related to this, I did find a dodge direction bug that results in backward flips not giving enough force when moving forwards, which I fixed. However, dodge torque was untouched in that change.

Perhaps this is an issue with bindings, or my new changes magically fixed it? 

Even giving complete garbage orientation inputs, like super tiny numbers, numbers that are too big, and NaNs, doesn't seem to cause any NaN physics values.

![image](https://user-images.githubusercontent.com/36944229/218603233-152959d2-4cc2-48bb-8b10-771857c81a17.png)

**Comment by ZealanL** [comment]
Ok I managed to reproduce it, it seems all of the inputs are required:

![image](https://user-images.githubusercontent.com/36944229/218607195-2ad85cc0-4f4b-446c-bf9a-05dd160a0f38.png)

**Comment by ZealanL** [comment]
Fixed with https://github.com/ZealanL/RocketSim/commit/454713d3ac05f9df9c05eadb3bf681aff0fc0f11, `dodgeDir` was being normalized unsafely, which produced problems when the number `-0` was involved.

Thank you for reporting this! I somehow never encountered this bug in all my testing, weird.
There is also another bug I fixed involving dodge forces not being amplified correctly when flipping against your forward motion.
